src/main/java/tornadofx/Component.kt

```diff
@@ -155,110 +156,119 @@ abstract class Component : Configurable {
+            val fragment: T by lazy { find<T>(overrideScope, params) }
+            override fun getValue(thisRef: Component, property: KProperty<*>): T = fragment
+        }
 
     inline fun <reified T : Any> di(name: String? = null): ReadOnlyProperty<Component, T> = object : ReadOnlyProperty<Component, T> {
         var injected: T? = null
         override fun getValue(thisRef: Component, property: KProperty<*>): T {
-            val dicontainer = FX.dicontainer ?: throw AssertionError(
-                    "Injector is not configured, so bean of type ${T::class} cannot be resolved")
-            return dicontainer.let {
-                if (name != null) {
-                    it.getInstance<T>(name)
-                } else {
-                    it.getInstance()
-                }
-            }.also { injected = it }
+            val dicontainer = FX.dicontainer ?: throw AssertionError("Injector is not configured, so bean of type ${T::class} cannot be resolved")
+            return dicontainer.let { if (name != null) it.getInstance<T>(name) else it.getInstance() }.also { injected = it }
         }
     }
 
     val primaryStage: Stage get() = FX.getPrimaryStage(scope)!!
 
     // This is here for backwards compatibility. Removing it would require an import for the tornadofx.ui version
-    infix fun <T> Task<T>.ui(func: (T) -> Unit) = success(func)
+    infix fun <T> Task<T>.ui(func: (T) -> Unit): Task<T> = success(func)
 
     @Deprecated("Clashes with Region.background, so runAsync is a better name", ReplaceWith("runAsync"), DeprecationLevel.WARNING)
-    fun <T> background(func: FXTask<*>.() -> T) = task(func = func)
+    fun <T> background(func: FXTask<*>.() -> T): Task<T> = task(func = func)
 
     /**
      * Perform the given operation on an ScopedInstance of the specified type asynchronousyly.
      *
      * MyController::class.runAsync { functionOnMyController() } ui { processResultOnUiThread(it) }
      */
-    inline fun <reified T, R> KClass<T>.runAsync(noinline op: T.() -> R) where T : Component, T : ScopedInstance = task { op(find(scope)) }
+    inline fun <reified T, R> KClass<T>.runAsync(noinline op: T.() -> R): Task<R> where T : Component, T : ScopedInstance = task { op(find(scope)) }
 
     /**
      * Perform the given operation on an ScopedInstance class function member asynchronousyly.
      *
      * CustomerController::listContacts.runAsync(customerId) { processResultOnUiThread(it) }
      */
-    inline fun <reified InjectableType, reified ReturnType> KFunction1<InjectableType, ReturnType>.runAsync(noinline doOnUi: (ReturnType) -> Unit = {}): Task<ReturnType>
-            where InjectableType : Component, InjectableType : ScopedInstance = task { invoke(find(scope)) }.apply { ui(doOnUi) }
+    inline fun <reified InjectableType, reified ReturnType> KFunction1<InjectableType, ReturnType>.runAsync(
+        noinline doOnUi: (ReturnType) -> Unit = {}
+    ): Task<ReturnType> where InjectableType : Component,
```

Comment:
agreed that it should be broken up. I want to see another brake before or after  the =